src/cache/redis.rs

```diff
@@ -1,95 +1,102 @@
+        .await
+        .expect("Redis connection pool initialization failed")
+}
 
 pub struct ServiceCache(RedisPool);
 
-pub fn create_service_cache() -> ServiceCache {
-    ServiceCache(create_pool())
-}
-
-fn create_pool() -> RedisPool {
-    let client = redis::Client::open(redis_uri()).unwrap();
-    Pool::builder().max_size(15).build(client).unwrap()
+pub async fn create_service_cache() -> ServiceCache {
+    ServiceCache(create_pool().await)
 }
 
 impl ServiceCache {
-    fn conn(&self) -> RedisConnection {
-        self.0.get().unwrap()
+    async fn conn(&self) -> RedisConnection<'_> {
+        self.0.get().await.unwrap()
     }
 }
 
+#[rocket::async_trait]
 impl Cache for ServiceCache {
-    fn fetch(&self, id: &str) -> Option<String> {
-        match self.conn().get(id) {
+    async fn fetch(&self, id: &str) -> Option<String> {
+        match self.conn().await.get(id).await {
             Ok(value) => Some(value),
             _ => None,
         }
     }
 
-    fn create(&self, id: &str, dest: &str, timeout: usize) {
-        let _: () = self.conn().pset_ex(id, dest, timeout).unwrap();
+    async fn create(&self, id: &str, dest: &str, timeout: usize) {
+        let _: () = self.conn().await.pset_ex(id, dest, timeout).await.unwrap();
     }
 
-    fn insert_in_hash(&self, hash: &str, id: &str, dest: &str) {
-        let _: () = self.conn().hset(hash, id, dest).unwrap();
+    async fn insert_in_hash(&self, hash: &str, id: &str, dest: &str) {
+        let _: () = self.conn().await.hset(hash, id, dest).await.unwrap();
     }
 
-    fn get_from_hash(&self, hash: &str, id: &str) -> Option<String> {
-        self.conn().hget(hash, id).ok()
+    async fn get_from_hash(&self, hash: &str, id: &str) -> Option<String> {
+        self.conn().await.hget(hash, id).await.ok()
     }
 
-    fn has_key(&self, id: &str) -> bool {
-        let result: Option<usize> = self.conn().exists(id).ok();
+    async fn has_key(&self, id: &str) -> bool {
+        let result: Option<usize> = self.conn().await.exists(id).await.ok();
         result.map(|it| it != 0).unwrap_or(false)
     }
 
-    fn expire_entity(&self, id: &str, timeout: usize) {
-        let _: () = self.conn().pexpire(id, timeout).unwrap();
+    async fn expire_entity(&self, id: &str, timeout: usize) {
+        let _: () = self.conn().await.pexpire(id, timeout).await.unwrap();
     }
 
-    fn invalidate_pattern(&self, pattern: &str) {
-        pipeline_delete(
-            &mut self.conn(),
-            scan_match_count(&mut self.conn(), pattern, redis_scan_count()),
-        );
+    async fn invalidate_pattern(&self, pattern: &str) {
+        let mut con = self.conn().await;
+        let keys_cmd = scan_match_count_cmd(pattern, redis_scan_count());
+        let mut keys = keys_cmd.iter_async(&mut *con).await.unwrap();
+        pipeline_delete(&mut keys)
+            .await
+            .query_async::<_, ()>(&mut *con)
```

Comment:
Should we move this into the `pipeline_delete`?

Reply:
I explained in this [comment](https://github.com/gnosis/safe-client-gateway/pull/811#discussion_r812624179) why I took it out. Just as for that comment, if you prefer the approach where the mutable reference to the connection is passed, I can try investigating further.

Reply:
I think it is fair for now. I was more thinking about reusability. Currently you would have to repeat more code than was necessary before.